### PR TITLE
add alpha_dropout in nn.functional and nn.layer, test=develop

### PR DESCRIPTION
**Pull request by huangjun12**
### PR types
New features

### PR changes
APIs

### Describe
add apis of alpha_dropout:
(1) paddle.nn.functional.alpha_dropout(x, p=0.5, training=True, name = None)
(2) paddle.nn.AlphaDropout(p=0.5, name = None)

### CODE REVIEW

File: python/paddle/nn/functional/common.py [hunk 1]
```diff
@@ -446,3 +452,80 @@ def _is_list_or_turple_(data):
...
+        p = layers.fill_constant(shape=[1], dtype='float32', value=p)
+        keep_mask = layers.greater_equal(random_tensor, p)
+        keep_mask = layers.cast(keep_mask, dtype)
+        drop_mask = layers.elementwise_sub(
+            layers.fill_constant(
+                shape=input_shape, dtype=dtype, value=1.),
+            keep_mask)
+
+        #apply mask
+        alpha_p = layers.fill_constant(shape=[1], dtype=dtype, value=alpha_p)
```
Comment (iclementine) [inline python/paddle/nn/functional/common.py]: 这些 tensor * scalar 的情形其实用 scale 可能会更快。

File: python/paddle/nn/layer/common.py [hunk 2]
```diff
@@ -342,3 +343,59 @@ def forward(self, input):
             mode=self._mode,
             pad_value=self._pad_value,
             data_format=self._data_format)
+
+
+class AlphaDropout(layers.Layer):
+    """
+        :alias_main: paddle.nn.AlphaDropout
```
Comment (iclementine) [inline python/paddle/nn/layer/common.py]: alis 可以不用写了。由官网解析自动处理

File: python/paddle/nn/layer/common.py [hunk 3]
```diff
@@ -342,3 +343,59 @@ def forward(self, input):
...
+    Alpha Dropout is a type of Dropout that maintains the self-normalizing property. For an input with
+    zero mean and unit standard deviation, the output of Alpha Dropout maintains the original mean and
+    standard deviation of the input. Alpha Dropout fits well to SELU activate function by randomly setting
+    activations to the negative saturation value.
+
+    For more information, please refer to:
+    [Self-Normalizing Neural Networks](https://arxiv.org/abs/1706.02515)
+    `paddle.nn.functional.alpha_dropout`
+
+    In dygraph mode, please use ``eval()`` to indicate whether it is in test phrase or not.
```
Comment (iclementine) [inline python/paddle/nn/layer/common.py]: to switch to evaluation mode, where dropout is disabled.

**Review by iclementine — DISMISSED** [review]
LGTM

File: python/paddle/nn/functional/common.py (hunk 1 above)
Comment (huangjun12) [reply python/paddle/nn/functional/common.py]: Done

File: python/paddle/nn/layer/common.py (hunk 2 above)
Comment (huangjun12) [reply python/paddle/nn/layer/common.py]: Done

File: python/paddle/nn/layer/common.py (hunk 3 above)
Comment (huangjun12) [reply python/paddle/nn/layer/common.py]: Done

File: python/paddle/nn/functional/common.py [hunk 4]
```diff
@@ -456,6 +460,80 @@ def _is_list_or_turple_(data):
     return out
 
 
+def alpha_dropout(x, p=0.5, training=True, name=None):
+    """
+    alpha_dropout function.
```
Comment (willthefrog) [inline python/paddle/nn/functional/common.py]: delete this line

File: python/paddle/nn/functional/common.py [hunk 5]
```diff
@@ -456,6 +460,80 @@ def _is_list_or_turple_(data):
...
+    Alpha Dropout is a type of Dropout that maintains the self-normalizing property. For an input with
+    zero mean and unit standard deviation, the output of Alpha Dropout maintains the original mean and
+    standard deviation of the input. Alpha Dropout fits well to SELU activate function by randomly setting
+    activations to the negative saturation value.
+
+    Args:
+        x (Tensor): The input tensor. The data type is float32 or float64.
+        p (float): Probability of setting units to zero. Default 0.5.
+        training (bool): A flag indicating whether it is in train phrase or not. Default True.
+        name (str|None): A name for this layer(optional). If set None, the layer
```
Comment (willthefrog) [inline python/paddle/nn/functional/common.py]: use template phrase for name.

File: python/paddle/nn/functional/common.py [hunk 6]
```diff
@@ -456,6 +460,80 @@ def _is_list_or_turple_(data):
...
+                         will be named automatically.
+    Returns:
+        A Tensor representing the dropout, has same shape and data type with `x`.
```
Comment (willthefrog) [inline python/paddle/nn/functional/common.py]: with -> as

File: python/paddle/nn/functional/common.py [hunk 7]
```diff
@@ -456,6 +460,80 @@ def _is_list_or_turple_(data):
...
+    Examples:
+        .. code-block:: python
+            import paddle
+            import numpy as np
+
+            paddle.disable_static()
+            x = np.array([[-1, 1], [-1, 1]]).astype('float32')
+            x = paddle.to_tensor(x)
+            y_train = paddle.nn.functional.alpha_dropout(x, 0.5)
+            y_test = paddle.nn.functional.alpha_dropout(x, 0.5, training=False) #test
```
Comment (willthefrog) [inline python/paddle/nn/functional/common.py]: remove `#test`

File: python/paddle/nn/functional/common.py [hunk 8]
```diff
@@ -456,6 +460,80 @@ def _is_list_or_turple_(data):
...
+            print(x.numpy())
+            print(y_train.numpy())
+            # [[-0.10721093, 1.6655989 ], [-0.7791938, -0.7791938]] (randomly)
+            print(y_test.numpy())
+    """
+    if not isinstance(p, (float, int)):
```
Comment (willthefrog) [inline python/paddle/nn/functional/common.py:829]: docstring only says float

File: python/paddle/nn/layer/common.py [hunk 9]
```diff
@@ -348,6 +337,59 @@ def forward(self, input):
...
+            print(x.numpy())
+            print(y_train.numpy())
+            # [[-0.10721093, 1.6655989 ], [-0.7791938, -0.7791938]] (randomly)
+            print(y_test.numpy())
+   """
+
+    def __init__(self, p=0.5, name=None):
+        super(AlphaDropout, self).__init__()
+        self.p = p
+        self.training = _dygraph_tracer()._train_mode
```
Comment (willthefrog) [inline python/paddle/nn/layer/common.py]: `Layer` now has `train()`/`eval()`, this can be removed

**Review by willthefrog — CHANGES_REQUESTED** [review]

File: python/paddle/nn/functional/common.py (hunk 4 above)
Comment (huangjun12) [reply python/paddle/nn/functional/common.py]: Done

File: python/paddle/nn/functional/common.py (hunk 8 above)
Comment (huangjun12) [reply python/paddle/nn/functional/common.py:829]: Done

File: python/paddle/nn/layer/common.py (hunk 9 above)
Comment (huangjun12) [reply python/paddle/nn/layer/common.py]: Done

**Comment by huangjun12** [comment]
add alpha_dropout and AlphaDropout; 
following #26111 
(1) refine doc of dropout series
(2) remove _dygraph_tracer()._train_mode in Dropout and Dropout3D

File: python/paddle/nn/functional/common.py [hunk 10]
```diff
@@ -80,42 +80,34 @@ def interpolate(input,
         'trilinear' : Trilinear interpolation
         'nearest' : Nearest neighbor interpolation
         'bicubic' : Bicubic interpolation
+    Linear interpolation is the method of using a line connecting two known quantities
```
Comment (willthefrog) [inline python/paddle/nn/functional/common.py]: unrelated to this PR?

File: python/paddle/nn/functional/common.py [hunk 11]
```diff
@@ -181,28 +170,27 @@ def interpolate(input,
               D_out = D_{in} * scale_{factor}
               H_out = H_{in} * scale_{factor}
               W_out = W_{in} * scale_{factor}
-
     For details of linear interpolation, please refer to Wikipedia:
     https://en.wikipedia.org/wiki/Linear_interpolation.
-    
+
```
Comment (willthefrog) [inline python/paddle/nn/functional/common.py]: auto formatting?

File: python/paddle/nn/functional/common.py [hunk 12]
```diff
@@ -459,12 +443,320 @@ def _is_list_or_turple_(data):
...
+                [[1 0 1]]
+                Doing elementwise multiply the mask will be broadcast from 1*3 to 2*3:
+                [[1 0 1]
+                 [1 0 1]]
+                and the result after elementwise multiply is:
+                [[1 0 3]
+                 [4 0 6]]
+            (3) What about ``axis=[0, 1]`` ? This means the dropout is performed in all axes of x,
+                which is the same case as default setting ``axis=None`` .
+            (4) You may note that logically `axis=None` means the dropout is performed in no axis of x,
```
Comment (willthefrog) [inline python/paddle/nn/functional/common.py]: `no axis` -> `all axis`

File: python/paddle/nn/functional/common.py [hunk 13]
```diff
@@ -459,12 +443,320 @@ def _is_list_or_turple_(data):
...
+                [[0 0 0]
+                 [0 0 0]]
+                and the result after elementwise multiply is:
+                [[0 0 0]
+                 [0 0 0]]
+                Actually this is not what we want because all elements may set to zero~
+            When x is a 4d tensor with shape `NCHW`, we can set ``axis=[0,1]`` and the dropout will be performed
+            in channel `N` and `C`, `H` and `W` is tied, i.e.
+            paddle.nn.dropout(x, p, axis=[0,1])
+            This is something we called dropout2d. Please refer to ``paddle.nn.functional.dropout2d``
```
Comment (willthefrog) [inline python/paddle/nn/functional/common.py]: remove `This is something we called dropout2d.`

File: python/paddle/nn/functional/common.py (hunk 11 above)
Comment (huangjun12) [reply python/paddle/nn/functional/common.py]: Done

File: python/paddle/nn/functional/common.py (hunk 13 above)
Comment (huangjun12) [reply python/paddle/nn/functional/common.py]: Done

**Review by TCChenlong — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by willthefrog — APPROVED** [review]
LGTM
